src/gui/plugins/apply_force_torque/ApplyForceTorque.cc

```diff
@@ -398,5 +566,343 @@ void ApplyForceTorquePrivate::PublishWrench(bool _applyForce, bool _applyTorque)
+    double scale = applicationPoint.Distance(this->camera->WorldPose().Pos())
+      / 2.0;
+    this->wrenchVis.UpdateVectorVisual(
+      this->forceVisual, worldForce, applicationPoint, scale, true);
+  }
+  else
+  {
+    this->forceVisual->SetVisible(false);
+  }
+
+  // Update torque visualization
+  if (this->torque != math::Vector3d::Zero &&
+      this->selectedEntity.has_value())
+  {
+    math::Vector3d worldTorque =
+      this->linkWorldPose.Rot().RotateVector(this->torque);
+    math::Vector3d applicationPoint =
+      this->linkWorldPose.Pos() + this->inertialPos;
+    double scale = applicationPoint.Distance(this->camera->WorldPose().Pos())
+      / 2.0;
+    this->wrenchVis.UpdateVectorVisual(
+      this->torqueVisual, worldTorque, applicationPoint, scale, false);
+  }
+  else
+  {
+    this->torqueVisual->SetVisible(false);
+  }
+
+  // Update gizmo visual
+  if (this->vector != RotationToolVector::NONE)
+  {
+    math::Vector3d pos;
+    math::Vector3d u;
+    if (this->vector == RotationToolVector::FORCE
+        && this->force != math::Vector3d::Zero)
+    {
+      pos =
+        this->linkWorldPose.Pos() + this->inertialPos +
+        this->linkWorldPose.Rot().RotateVector(this->offset);
+      u = this->force;
+    }
+    else if (this->vector == RotationToolVector::TORQUE
+            && this->torque != math::Vector3d::Zero)
+    {
+      pos = this->linkWorldPose.Pos() + this->inertialPos;
+      u = this->torque;
+    }
+    else
+    {
+      this->vector = RotationToolVector::NONE;
+      this->gizmoVisual->SetTransformMode(rendering::TransformMode::TM_NONE);
+      this->gizmoVisual->SetActiveAxis(math::Vector3d::Zero);
+      return;
+    }
+
+    double scale = pos.Distance(this->camera->WorldPose().Pos())
+      / 2.0;
+
+    // Update gizmo visual rotation so that they are always facing the
+    // eye position
+    math::Quaterniond gizmoRot = this->linkWorldPose.Rot() * this->vectorRot;
+    math::Vector3d eye = gizmoRot.RotateVectorReverse(
+      (this->camera->WorldPosition() - pos).Normalized());
+
+    rendering::VisualPtr xVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_X);
+    xVis->SetVisible(false);
+
+    rendering::VisualPtr yVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_Y);
+    math::Vector3d yRot(0, atan2(eye.X(), eye.Z()), 0);
+    math::Vector3d yRotOffset(GZ_PI * 0.5, -GZ_PI * 0.5, 0);
+    yVis->SetWorldRotation(gizmoRot *
+      math::Quaterniond(yRot) * math::Quaterniond(yRotOffset));
+
+    rendering::VisualPtr zVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_Z);
+    math::Vector3d zRot(0, 0, atan2(eye.Y(), eye.X()));
+    zVis->SetWorldRotation(gizmoRot * math::Quaterniond(zRot));
+
+    rendering::VisualPtr circleVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_Z << 1);
```

Comment:
FYI: https://github.com/gazebosim/gz-rendering/issues/880

Reply:
This looks like same logic in `TransformController` class in gz-rendering. I'll provide a function in `GizmoVisual` class in gz-rendering to take care of this logic so we can avoid duplicating the code here. The changes are ABI breaking. So I would suggest leaving the changes like this in gz-sim7 (maybe with a todo note) and we can remove the code in `main`

Reply:
Sounds good!

Reply:
https://github.com/gazebosim/gz-rendering/pull/882